### PR TITLE
[RFR] Add oauth related models

### PR DESCRIPTION
**Pull request by cwisecarver**

### CODE REVIEW

**Comment by cwisecarver** [comment]
Covers #16 #17 and ApiOAuth2Scope, which was apparently not a story, which is terrifying.

File: osf_models/models/oauth.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    This collection is also used by CAS to create the master list of available tokens.
+    Any changes made to field names in this model must be echoed in the CAS implementation.
+    """
+
+    # Name of the field being `token_id` is a CAS requirement.
+    # This is the actual value of the token that's used to authenticate
+    token_id = models.CharField(max_length=70, default=functools.partial(random_string, length=70),
+                                  unique=True)
+
+    owner = models.ForeignKey('OSFUser', db_index=True, blank=False, null=False)
```
Comment (sloria) [inline osf_models/models/oauth.py]: Minor: in future versions, `on_delete` will become a required argument. I think it should be set to `SET_NULL` here, yes?

File: osf_models/models/oauth.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    client_id = models.UUIDField(default=uuid.uuid4,  # Not *guaranteed* unique, but very unlikely
+                                   unique=True,
+                                   db_index=True)
+
+    client_secret = models.CharField(default=generate_client_secret, max_length=40)
+
+    is_active = models.BooleanField(default=True,  # Set to False if application is deactivated
+                                    db_index=True)
+
+    owner = models.ForeignKey('OSFUser', null=False, blank=False)
```
Comment (sloria) [inline osf_models/models/oauth.py]: Same here: `on_delete=models.SET_NULL`.

File: osf_models/models/oauth.py (hunk 1 above)
Comment (cwisecarver) [inline osf_models/models/oauth.py]: Correct. I’ll fix it and start using it from now on.

File: osf_models/models/oauth.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    # User-specified application descriptors
+    name = models.CharField(db_index=True, blank=False, null=False, max_length=200)
+    description = models.CharField(blank=True, null=True, max_length=1000)
+
+    date_created = models.DateTimeField(auto_now_add=True)
```
Comment (sloria) [inline osf_models/models/oauth.py]: Won't this need to be `default=timezone.now` for the migration's sake?

**Comment by sloria** [comment]
Just a few minor comments above. Once those are addressed, go ahead and merge.

**Pull request merged**